### PR TITLE
Move TLS options to scrape config

### LINKED ISSUE #989: Standardize auth configuration/handling for different SD methods
The Kubernetes service discovery added support for querying the Kubernetes API with different authentication mechanisms. This is also useful for other types of service discovery but should be configurable in a common way.

### LINKED ISSUE #1013: Insecure certificate
Hi,

In my case i didn't have a valid certificate for my domain so when i try to get metrics with https on my openshift I have : 

```bash
Get https://10.0.0.254:8443/metrics: x509: cannot validate certificate for 10.0.0.254 because it doesn't contain any IP SANs 
```
In the kubernetes_sd_configs, we have the insecure flag, but i didn't find in the scrape_config part
Is it possible to have something like that ?

```yaml
  - job_name: 'openshift-metrics'
    scheme: https
    insecure: True

    target_groups:
      - targets: ['10.0.0.254:8443']

    bearer_token: '<token>'
```

Thanks,

### PR DESCRIPTION
**Pull request by jimmidyson**
Fixes #1013, fixes #989

### CODE REVIEW

**Comment by jimmidyson** [comment]
Is that failure just a flake or something I need to fix?

**Comment by fabxc** [comment]
Flake – something that needs to be fixed. But not by you, and not in this PR :)

**Comment by jimmidyson** [comment]
@fabxc Right answer!

**Comment by fabxc** [comment]
Don't we have duplication (with different naming) now between scrape config and k8s config?
Would a dedicated config block make sense in general? 

```
tls_config: {
	insecure   bool
	cert_file  string
	key_file   string
	ca_file    string
}
```

Especially in the scrape_config this seems to make sense. With that many options the `insecure` field kind of means all and nothing.

**Comment by jimmidyson** [comment]
@fabxc Sounds good. Inlined the type in config?

**Comment by fabxc** [comment]
I would just make it separate so we can use the type in tests, attach methods etc.

**Comment by jimmidyson** [comment]
@fabxc Updated.

File: config/config.go [hunk 1]
```diff
@@ -315,12 +332,10 @@ type ScrapeConfig struct {
...
 	// The bearer token file for the targets.
 	BearerTokenFile string `yaml:"bearer_token_file,omitempty"`
-	// The ca cert to use for the targets.
-	CACert string `yaml:"ca_cert,omitempty"`
-	// The client cert authentication credentials for the targets.
-	ClientCert *ClientCert `yaml:"client_cert,omitempty"`
 	// HTTP proxy server to use to connect to the targets.
 	ProxyURL URL `yaml:"proxy_url,omitempty"`
+	// Inlined TLSConfig.
+	TLSConfig TLSConfig `yaml:",inline"`
```
Comment (fabxc) [inline config/config.go]: Using a separate struct makes this now clearer for the programmer. The inlining still puts all the attributes onto the same level for the user, which is what I was most concerned about.

Why do you preferring it this way?

@juliusv opinion?

**Comment by fabxc** [comment]
Generally looks good.

File: config/config.go (hunk 1 above)
Comment (brian-brazil) [inline config/config.go]: From a quick look at the current tls.Config there's another 8 options we may need to add. I don't think what's potentially a 13 field struct is sane to inline, especially as some of the names don't make it obvious that they're part of SSL configuration (e.g. insecure, servername, minversion, maxversion, nextprotos).

I'm likely going to end up using this code over in the blackbox exporter too,

Comment (juliusv) [inline config/config.go]: Yeah, :+1: for having them not inlined, but in their own subsection. Already things like `KeyFile` and `Insecure` could mean anything if it's not clear that they are about TLS.

File: config/config.go [hunk 2]
```diff
@@ -293,6 +298,18 @@ func (c *GlobalConfig) isZero() bool {
 		c.EvaluationInterval == 0
 }
 
+// TLSConfig configures the options for TLS connections.
+type TLSConfig struct {
+	// The ca cert to use for the targets.
```
Comment (brian-brazil) [inline config/config.go]: CA, as it's an acronym

File: config/config.go (hunk 1 above)
Comment (brian-brazil) [inline config/config.go]: `Insecure` is also a bit unclear, I'd go with the upstream name of `InsecureSkipVerify` if we're changing this anyway.

Comment (juliusv) [inline config/config.go]: Or just `AllowInsecure` or `SkipVerification`?

Comment (brian-brazil) [inline config/config.go]: As a general rule I think we should probably stick with upstream naming unless we've got a good reason not to. `SkipCertVerification` would be clearer still, but I'm not sure it's sufficiently better to add an extra level of indirection for users trying to figure out what the option does.

Comment (juliusv) [inline config/config.go]: But most of our users don't have any expectation that our configuration field names have any relationship to internal Go standard library option names (or would know those names). So I see the benefits of allowing different names more than the downsides, and `InsecureSkipVerify` is not the clearest name, right?

Comment (brian-brazil) [inline config/config.go]: If I saw a config option that I was unsure about I'd Google it, and the go docs in the case are the best ones to explain what that option does. If it's named differently I'd have to dig through the code only to find out it's the same option, but a different name.

Comment (fabxc) [inline config/config.go]: I agree that a mapping to Go internals is somewhat preferable. Users who actually make a debugging effort themselves would take that route.

Comment (jimmidyson) [inline config/config.go]: Comments addressed:

* Moved to `InsecureSkipVerify` - horrible name but agree it makes more sense to sync with go libs
* `TLSConfig` is now a nested struct for the reasons discussed above.

File: config/config.go [hunk 3]
```diff
@@ -293,6 +298,18 @@ func (c *GlobalConfig) isZero() bool {
...
+// TLSConfig configures the options for TLS connections.
+type TLSConfig struct {
+	// The CA cert to use for the targets.
+	CAFile string `yaml:"ca_file,omitempty"`
+	// The client cert file for the targets.
+	CertFile string `yaml:"cert_file,omitempty"`
+	// The client key file for the targets.
+	KeyFile string `yaml:"key_file,omitempty"`
+	// Disable target certificate validation.
+	InsecureSkipVerify bool `yaml:"insecure_skip_verify,omitempty"`
```
Comment (fabxc) [inline config/config.go]: Always seeing this clearly is better, let's drop the `omitempty`.

**Comment by fabxc** [comment]
:+1: after that

File: config/config.go (hunk 3 above)
Comment (jimmidyson) [inline config/config.go]: Done.

File: cmd/promtool/main.go [hunk 4]
```diff
@@ -108,13 +108,18 @@ func checkConfig(t cli.Term, filename string) ([]string, error) {
...
+		}
+		if err := check(scfg.TLSConfig.KeyFile); err != nil {
+			return nil, fmt.Errorf("error checking client key file %q: %s", scfg.TLSConfig.KeyFile, err)
+		}
+
+		if len(scfg.TLSConfig.CertFile) > 0 && len(scfg.TLSConfig.KeyFile) == 0 {
+			return nil, fmt.Errorf("client cert file %s specified without client key file", scfg.TLSConfig.CertFile)
+		}
+		if len(scfg.TLSConfig.KeyFile) > 0 && len(scfg.TLSConfig.CertFile) == 0 {
+			return nil, fmt.Errorf("client key file %s specified without client cert file", scfg.TLSConfig.KeyFile)
```
Comment (juliusv) [inline cmd/promtool/main.go]: Aren't the same checks also needed for the TLSConfig contained in the KubernetesSDConfig?

Comment (jimmidyson) [inline cmd/promtool/main.go]: Yes they are - must have missed that in the refactoring. Sorry!

File: config/config.go [hunk 5]
```diff
@@ -315,12 +332,10 @@ type ScrapeConfig struct {
 	BearerToken string `yaml:"bearer_token,omitempty"`
 	// The bearer token file for the targets.
 	BearerTokenFile string `yaml:"bearer_token_file,omitempty"`
-	// The ca cert to use for the targets.
-	CACert string `yaml:"ca_cert,omitempty"`
-	// The client cert authentication credentials for the targets.
-	ClientCert *ClientCert `yaml:"client_cert,omitempty"`
 	// HTTP proxy server to use to connect to the targets.
 	ProxyURL URL `yaml:"proxy_url,omitempty"`
+	// Inlined TLSConfig.
```
Comment (juliusv) [inline config/config.go]: This isn't inlined anymore, so adjust the comment?

Comment (jimmidyson) [inline config/config.go]: I blame the reviewer(s) ;)

Comment (fabxc) [inline config/config.go]: I take the blame :)

File: cmd/promtool/main.go (hunk 4 above)
Comment (jimmidyson) [inline cmd/promtool/main.go]: Cleaned up in #1063

**Pull request merged**